### PR TITLE
feat(cli): Local federation development mode

### PR DESCRIPTION
**Pull request by pimeys**
Adds local development mode for federated graphs.

To start the server:

```
gb dev --federated
```

This doesn't need to be in a project (yet), but that will change this week.

Then go into a project which is a subgraph, start the dev in there

```
gb dev
```

And finally publish the subgraph to the federated graph:

```
gb publish --dev --name one --url "http://localhost:4001/graphql" -H "x-api-key: kekw"
```

The URL should be the one which answers to GraphQL queries in the subgraph. It'll introspect the schema and compose it to the subgraph.

Closes: GB-5343

### CODE REVIEW

File: cli/crates/federated-dev/src/dev/composer/messages.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (obmarg) [inline cli/crates/federated-dev/src/dev/composer/messages.rs:1]: Is this file meant to be here?

File: cli/crates/federated-dev/src/dev/admin/types.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#[derive(InputObject)]
+pub(crate) struct PublishSubgraphInput {
+    pub(crate) name: String,
+    pub(crate) url: Url,
+    pub(crate) headers: Vec<Header>,
+}
+
+#[derive(SimpleObject)]
+pub(crate) struct PublishSubgraphSuccess {
+    __typename: &'static str,
```
Comment (obmarg) [inline cli/crates/federated-dev/src/dev/admin/types.rs:29]: Quite odd that you're defining this - `__typename` implicitly exists on all GraphQL composite outputs.

Could you just do `struct PublishSubgraphSuccess` and save yourself some effort?

File: cli/crates/federated-dev/src/subgraph.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            url: "{url}",
+            headers: [{headers}]
+          }}) {{ typename }}
+        }}
+    "#};
+
+    let request = json!({
+        "query": mutation,
+        "variables": {}
+    });
```
[the remark this answers is not in the document]
Comment (obmarg) [reply cli/crates/federated-dev/src/subgraph.rs:35]: Oh, also - where's the `typename` field coming from?  It's defined as `__typename` in your struct.  (Imagine async-graphql might be stripping off the `__` so your object has both `typename` (the field you defined) _and_ `__typename` (the implicit field).

Might be worth cleaning up?

**Review by obmarg — APPROVED** [review]
Good stuff, just a few small comments

File: cli/crates/federated-dev/src/dev/composer.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    fn ingest_subgraphs(&self, add_new: Option<(&str, &Subgraph)>) -> Subgraphs {
+        let mut subgraphs = Subgraphs::default();
+
+        for (name, subgraph) in &self.graphs {
+            if add_new.is_some_and(|(added_name, _)| added_name == name) {
+                continue;
+            }
+
+            subgraphs.ingest(subgraph.schema(), name);
```
Comment (obmarg) [inline cli/crates/federated-dev/src/dev/composer.rs:94]: I just changed the signature of (I think) this function btw - sorry for the breakage.  Probably want to pull in `main` and add `subgraph.url()` (or whatever it is) on the end

File: cli/crates/federated-dev/src/dev/admin/types.rs (hunk 2 above)
Comment (pimeys) [reply cli/crates/federated-dev/src/dev/admin/types.rs:29]: async-graphql expects at least one field in the object. I don't really want to return anything if successful, so I wonder what we could have here?

Comment (obmarg) [reply cli/crates/federated-dev/src/dev/admin/types.rs]: Could just return a `boolean` or similar rather than a whole object?

Comment (pimeys) [reply cli/crates/federated-dev/src/dev/admin/types.rs]: Ah, it was late yesterday, of course...

File: cli/crates/cli/src/cli_input.rs [hunk 5]
```diff
@@ -340,24 +354,60 @@ pub struct SchemaCommand {
...
+    /// Add a header to the introspection request
+    #[clap(short = 'H', long, value_parser, num_args = 0..)]
+    header: Vec<String>,
+}
+
+impl PublishCommand {
+    pub(crate) fn headers(&self) -> impl Iterator<Item = (&str, &str)> {
+        self.header.iter().map(|header| {
+            let mut splitted = header.split(':');
+            (splitted.next().unwrap_or(""), splitted.next().unwrap_or("").trim())
```
[the remark this answers is not in the document]
Comment (pimeys) [reply cli/crates/cli/src/cli_input.rs:402]: Yeah I have an idea how to do this correctly. It kind of requires a proper parser, but luckily a proper parser here is _very_ simple.

**Pull request merged**